File: code/sorting/shaker_sort/shaker_sort.go
// Part of Cosmos by OpenGenus Foundation
package main

import "fmt"

func shakerSort(sortMe []int) {
	beginningIdx := 0
	endIdx := len(sortMe) - 1
	for beginningIdx <= endIdx {
		newBeginningIdx := endIdx
		newEndIdx := beginningIdx
		swap := false
		for i := beginningIdx; i < endIdx; i++ {
			if sortMe[i] > sortMe[i+1] {
				sortMe[i], sortMe[i+1] = sortMe[i+1], sortMe[i]
				swap = true
				newEndIdx = i
			}
		}
		if !swap {
			break
		}
		endIdx = newEndIdx
		swap = false
		for i := endIdx; i >= beginningIdx; i-- {
			if sortMe[i] > sortMe[i+1] {
				sortMe[i], sortMe[i+1] = sortMe[i+1], sortMe[i]
				swap = true
				newBeginningIdx = i
			}
		}
		if !swap {
			break
		}
		beginningIdx = newBeginningIdx
	}
}

func main() {
	sortMe := [10]int{6, 4, 9, 5, 3, 10, 8, 1, 7, 2}
	fmt.Println(sortMe)
	shakerSort(sortMe[0:10])
	fmt.Println(sortMe)
}

File: code/sorting/src/shaker_sort/shaker_sort.py
# Part of Cosmos by OpenGenus Foundation
import random


def shaker_sort(sort_list):
    beginning_idx = 0
    end_idx = len(sort_list) - 1
    while beginning_idx <= end_idx:
        swap = False
        for i in range(beginning_idx, end_idx):
            if sort_list[i] > sort_list[i + 1]:
                sort_list[i], sort_list[i + 1] = sort_list[i + 1], sort_list[i]
                swap = True
                end_idx = i
        if not swap:
            break
        swap = False
        for i in reversed(range(beginning_idx, end_idx)):
            if sort_list[i] > sort_list[i + 1]:
                sort_list[i], sort_list[i + 1] = sort_list[i + 1], sort_list[i]
                swap = True
                beginning_idx = i
        if not swap:
            break
    return sort_list


if __name__ == "__main__":
    count = 20

    source_list = []
    random.seed()
    for j in range(0, count):
        source_list.append(random.randint(0, 100))

    print("Input:", source_list)
    shaker_sort(source_list)
    print("Output:", source_list)
